[tooling] Add `dep_why` helper script to `dep_tree_resolver` tooling

sgnn7 opened the pull request:
### What does this PR do?

- Adds `dep_why.py` script which helps identify why a particular import in the
  dependency chain is pulled in.
- Adds a bit more info to the `dep_tree_resolver` README about caveats

### Motivation

While output of `dep_tree_resolver` is human-parsable, the file is usually too
large (100+ MB) to easily go through the file and identify the dependency chain
of the module of interest. This new script can be used to automate this process.

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

N/A

### Checklist

- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

sgnn7 committed:
[tooling] Improved docs for `dep_tree_resolver` tool

We needed a bit more exact caveats on this toos about not removing
dependencies from the output that are explicitly defined but end
up not being used by the parent module (and thus the agent).

sgnn7 committed:
[tooling] Add `dep_why.py` tool to help identify dependency links

This tool automates a common use case of the dep_tree_resolver tool
output by scanning the generated dependency tree and printing out any
tree branches that lead to the specified module name. This is useful in
identifying the reasons why a particular module is being used and
through which hierarchy of dependencies.

sgnn7 committed:
[docs] Add documentation about how to use `dep_why.py`

This should make it easier for people trying to use this tool.

mx-psi commented on tools/dep_tree_resolver/dep_why.py:
```diff
@@ -0,0 +1,67 @@
+    with open(dep_tree_filename) as infile:
+        for line in infile:
+            level, module_dep = parse_line(line)
+
+            if level <= len(current_path):
+                for _ in range(len(current_path) - level + 1):
+                    if len(current_path) > 0:
+                        current_path.pop()
+
```
I think this is equivalent, right?
```suggestion
                current_path = current_path[:level+1]
```

I find it easier to follow when written that way (but maybe that is just me)

sgnn7 replied:
Yeah that's much simpler and easier to follow! 

Fixed via https://github.com/DataDog/datadog-agent/pull/9308/commits/0985b2d57a3c059c25589cdf0fc0b56961eac6fd

sgnn7 committed:
Update tools/dep_tree_resolver/dep_why.py

Co-authored-by: Pablo Baeyens <pablo.baeyens@datadoghq.com>

sgnn7 committed:
Update tools/dep_tree_resolver/README.md

Co-authored-by: Pablo Baeyens <pablo.baeyens@datadoghq.com>

sgnn7 committed:
[tools] Make dep_why use a simpler hierarchy level calculation

Small change to optimize the level reduction as we go through the
hierarchy.

mx-psi commented on tools/dep_tree_resolver/dep_why.py:
```diff
@@ -0,0 +1,65 @@
+        for line in infile:
+            level, module_dep = parse_line(line)
+
+            if level <= len(current_path):
+                current_path = current_path[: level + 1]
```
You don't need the `len` check:
```python3
❯ python3
# ...
>>> a = [1,2,3,4,5]
>>> a[:4]
[1, 2, 3, 4]
>>> a[:200]
[1, 2, 3, 4, 5]
>>>
```

mx-psi reviewed (approved):
Looks good except for the nit above :) You will need to rebase to get all the pipeline fixes from the last days

sgnn7 committed:
[tools] Remove unnecessary logic branch in `dep_why`

The special case is no longer needed to check the length since array
bounds work fine as-is even with out-of-bounds indexing.

sgnn7 committed:
Merge remote-tracking branch 'origin/main' into sgnn7/add-dep-why-tooling
